### ISSUE TITLE
Pod Termination handling kicks in before the load balancer has had time to remove target

### ISSUE THREAD

**Issue by nirnanaaa** [kubernetes/kubernetes#106476]
### What happened?

When a pod is entering its Terminating state, it will receive a signal, asking it kindly to finish up work after which kubernetes will proceed deleting the pod. 

At the same time that the pod starts terminating, an ingress controller will receive the updated endpoints object, which will start removing the pod from the list of targets in the load balancer, that traffic could be sent to.

Both of these processes - the signal handling at the kubelet level and the removal of the Pods IP from the list of endpoints - are decoupled from one another and the SIGTERM might have been handled before, or at the same time, that the target in the target group is being processed.

As result the ingress controller might still send traffic to targets, which are still in its endpoints, but have properly shut down already. This might result in dropped connections, as the LB is still trying to send requests to the properly shutdown pod. The LB will in-turn reply with 5xx responses.

### What did you expect to happen?

no traffic being dropped during shutdown.

The SIGTERM should only start after the ingress controller/LB has removed the target from the target group. Readiness gates work pretty good for pod startup/rollout but lack support during pod deletion.

### How can we reproduce it (as minimally and precisely as possible)?

This is a very theoretical problem, which is very hard to reproduce:

- Provision an ingress controller (AWS LB for example)
- Create an ingress
- Create a service and pods (multiple ones through a deployment work best) for this ingress
- (add some delay/load to the cluster, that will cause the LB synchronization to be slower or delayed)
- startup an HTTP benchmark to produce some artificial load
- rollout a change to the deployment or just evict some pods

### Anything else we need to know?

We've been relying on [Pod-Graceful-Drain](https://github.com/foriequal0/pod-graceful-drain), which unfortunately intercepts and breaks k8s internals.

You can achieve a pretty good result as well using a `sleep` as `preStop`, but that's not reliable at all - due to the fact that it's just a guessing game if your traffic will be drained after X seconds - and requires statically linked binaries to be mounted in each container or the existence of sleep in the operating system.

I also opened up an issue on the [Ingress Controllers repo](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2366). 

### Kubernetes version

<details>

```console
$ kubectl version
v1.18.20
```

</details>

### Cloud provider

<details>
AWS/EKS
</details>

### OS version

<details>

```console
# On Linux:
sh-4.2$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"

# paste output here
$ uname -a
Linux xxx 4.14.252-195.483.amzn2.x86_64 #1 SMP Mon Nov 1 20:58:46 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>
- [https://github.com/kubernetes-sigs/aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller)
</details>

### Container runtime (CRI) and and version (if applicable)

<details>
Docker version 20.10.7, build f0df350
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by uablrek** [comment]
Blog that describes the problem at; https://medium.com/flant-com/kubernetes-graceful-shutdown-nginx-php-fpm-d5ab266963c2

I especially like the chart. It illuminates the problem in K8s; there is no kube-proxy<->kubelet communication (and there shouldn't be!). They both react independently on kube-api updates.

**Comment by nirnanaaa** [comment]
Indeed. (the part in their article about sigterm is very missleading, because it doesn't matter how well your app handles sigterm, if new traffic will still arrive after your app shutdown was completed, that's the "Practice. Potential problems with graceful shutdown" in their article).

I totally get that there is not supposed to be a link, don't get me wrong, most of these errors could - and probably should - be solved on client side, but unfortunately it's very hard for users to understand the behavior and mechanics. They're just used to these kinds of dependencies.

I just feel a `sleep` is not a guarantee. Even though it might be the most simple approach to solving a majority of the cases, it drives people crazy, who are trying to figure out the remaining errors.

I do think the mechanic of `preStop` is the right way to approach this. But with a more event-based release. wdyt?

**Comment by uablrek** [comment]
I don't think this can be done "by the means of kube-api" so to speak. A pod can get closer by monitoring it's own endpoints, but beside being quite complex it will still not know if all kube-proxys has updated the actual load-balancing. To introduce some status for (removed) endpoints that is updated when all kube-proxy instances has updated load-balancing, I find horrible. Just imagine the cluster wide sync and all possible fault cases :anguished:

IMO this falls into the "service mesh" domain. They monitor actual connections I think (circuit breaking?), but I am not very familiar with service meshs I must admit.

**Comment by andrewsykim** [comment]
EndpointSlice API now supports terminating condition, wonder if ingress controllers can be updated to leverage this for graceful termination of endpoints?

**Comment by thockin** [comment]
As @uablrek described, syncing state from all kube-proxies on every change to every endpoint of every service isn't feasible.  I'm not sure terminating endpoints helps much, if the core problem is that the Ingress controller or proxy is out-to-lunch or otherwise not getting updates.

**Comment by nirnanaaa** [comment]
I feel that the actual problem is that pods just get their termination signals way too early in this process, right ? Not sure what ingress controllers or kube proxy should do about that.

**Comment by aojea** [comment]
> I feel that the actual problem is that pods just get their termination signals way too early in this process, right ? Not sure what ingress controllers or kube proxy should do about that. If a container process is down it doesn’t matter if kube-proxy or the controller stops sending traffic there eventually - all traffic in between these events will still end up hitting an already dead target.

the process is not parallel is sequential and async

kill pod -> pod goes not ready -> endpoint controller receives event pod is not ready -> updates the endpoints -> ingress receive and event endpoints has changed -> ...

**Comment by nirnanaaa** [comment]
true, but shouldn't the pod be able stay alive until that event has been propagated and processed (just as a preStop hook kinda does)? This would ensure that the container doesn't even enter sigterm until kube-proxy/$controller had time to remove the endpoint.

**Comment by aojea** [comment]
> true, but shouldn't the pod be able stay alive until that event has been propagated and processed (just as a preStop hook kinda does)? This would ensure that the container doesn't even enter sigterm until kube-proxy/$controller had time to remove the endpoint.

how do the pod know that?

> syncing state from all kube-proxies on every change to every endpoint of every service isn't feasible.

I see your point, you want to make the process completely synchronous, but is not how Kubernetes works https://kubernetes.io/docs/concepts/architecture/controller/#controller-pattern , oversimplifying you have a bunch of controllers that sync the current state to the desired state, and are eventually consistent,

**Comment by andrewsykim** [comment]
(fixing https://github.com/kubernetes/kubernetes/issues/106476#issuecomment-1016907646)

/triage accepted

**Comment by mbyio** [comment]
I've spent a lot of time debugging this issue in AWS. One particularly problematic use case is with network load balancers. They always take at least 2 minutes to deregister a pod in IP mode from when they receive a request via AWS' API to remove the pod. During this time, the NLB will still send new TCP connections to the pod.

The AWS Load Balancer controller would be able to detect when a pod is actually completely deregistered. So would other ingress controllers. So, I would love if we had a "termination gate" or similar, which could be added by a controller, similar to readiness gates. Actually, the AWS Load balancer controller already makes use of readiness gates, because NLB/ALB registration is just as slow. It's weird that Kubernetes doesn't do anything to help with deregistration as well.

If we had a termination gate, this would fix the problem, because we would complete the loop: k8s would say it wants to terminate a pod, controllers would have time to update and stop sending traffic, and then the pod would be told "okay, it's time to cleanup".

**Comment by motoki317** [comment]
The issue is still relevant for many of us. Not being able to shutdown correctly without using a hack of guessing the time for `sleep` is not an ideal situation.

Related issues (in aws lb controller and ingress-nginx):
https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2366
https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2131
https://github.com/kubernetes/ingress-nginx/issues/6928
https://github.com/kubernetes/ingress-nginx/issues/6287

> If we had a termination gate, this would fix the problem, because we would complete the loop: k8s would say it wants to terminate a pod, controllers would have time to update and stop sending traffic, and then the pod would be told "okay, it's time to cleanup".

I agree with this.
Is anyone aware of a currently submitted proposal (KEPS?) of this "termination gate" feature or something similar?

**Comment by thockin** [comment]
No such KEP exists, to the best of my knowledge.  It would require either a whole new lifecycle phase (deleted but waiting) or perhaps a new preStop mechanism like "wait for upstream LBs", which then asks the question "how do I know which LBs to wait for?"

**Comment by thockin** [comment]
Not all ingress controllers run in a place that is accessible by the cluster.  But that doesn't solve the problem -- we still don't know who to wait for.   There's no rule that says there is only one ingress or gateway or even service LB.  Even for the same pod.  Do we wait for all of them?  For how long?  Once deletion has started, it cannot be stopped.  Components like kubelet and others can use the pod'd termination grace period to bound future waiting - are we going to violate that?

I am not saying we can't solve this, but I haven't found a "just do X" sort of answer.  I suspect the solution is going to be significantly more complicated.  Compatibility is hard.

**Comment by michelesr** [comment]
> we still don't know who to wait for.

Send a webhook to an endpoint, containing the information about the pod that is getting deleted. The webhook reads the information, runs its own business logic to check that everything is updated as it should be, and then it returns ok, and at that point, you can send the termination signal to the containers.

Don't take assumptions on what the webhook should do, just let the user choose where to send the webhook and proceed when a 200 response is returned? A timeout should be configurable so that you just don't wait forever if the endpoint hangs.

**Comment by thockin** [comment]
> Send a webhook to an endpoint

Where is that endpoint configured?  What if I have two different ingress controllers in the same cluster?  What if I have 2 different ingress objects?  What if the ingress controller is running in a place I can't reach from my cluster (e.g. can reach into clusters through a control-plane proxy, not the reverse)?   

What if that webhook takes 5 minutes to respond but the pod's termination grace period is 30 seconds?  Does kubelet go ahead and kill it anyway?  Or do we block the beginning of deletion for 5 minutes?  Is there any indication in the Pod object itself that this is happening?  Since deletion hasn't really started, can it be aborted?  What if I turn up a NEW loadbalancer in that time period?

Also, webhooks are notorious for failure modes.  What if it fails, do I orphan LBs?  This is the opposite of declarative.

Again, I am not saying we can't solve this, but I want to disavow you of the idea that it's simple, or I would need a much more robust set of answers to the above questions (these are just off the top of my head!) to convince me that is IS simple.

**Comment by michelesr** [comment]
> Where is that endpoint configured

Somewhere in the pod spec just like readiness gate.

> What if I have two different ingress controllers in the same cluster?

Let the user decide which hook to use, setting it in the pod spec.

>  What if the ingress controller is running in a place I can't reach from my cluster

It's a reasonable limitation not to being able to set a webhook for an endpoint you can't reach

> What if that webhook takes 5 minutes to respond but the pod's termination grace period is 30 seconds? 

Maybe treat the hook just like prestop, so account the time into the termination grace period and prompt the user to set a higher termination period. OR don't account for the time but then give the user ability to set a timeout for the hook, and then start counting against the termination grace period.

> I want to disavow you of the idea that it's simple

I never thought this is simple, is clearly a challenge to implement correctly. Kubernetes is a complex piece of software, and I appreciate there are a lot of challenges when introducing new features.

**Comment by thockin** [comment]
Let's not turn this issue into the KEP.  There are a LOT of things to consider, and I feel that we've only begun to enumerate them.

I agree that pod lifecycle WRT external LBs is under-developed.  I think the majority of the problem is there (pod lifecycle, SIG Node) and not in networking.  It just happens that networking is the thing that needs it.

I don't see any "easy" systemic solution.  I suspect it's POSSIBLE to have a pod-based sidecar thing which could make this work in a DIY way, but it would be delicate to maintain at scale and over time.

To actually fix this would require a KEP and an owner who is committed to working through all the issues, and I again feel obligated to warn potential candidates that this water is deeper than it appears.  It's not a case of me having a design in mind and just needing execution - I legitimately do not know the answer (though I am sure we could find one).

@tallclair and @SergeyKanzhelev just because we have talked about pod lifecyce issues before.  The way we handle termination makes this inherently a race.

**Comment by mbyio** [comment]
I think it is possible to make some prototype solutions that are only based on adjusting a container's preStop hook. I'm not aware of any ingress controllers that have tried it so far. It seems like that would be a good first step.

**Comment by zac-nixon** [comment]
Hello! I have an interest in fixing this problem. I am starting the initial steps of the KEP by getting this solution doc reviewed with sig-node.

https://docs.google.com/document/d/15hRxWGeXNXkYqjfZyemlEamMH5jiL7zzTsdbYqHWLdg/edit?usp=sharing

I'll post the KEP here if my proposal gets accepted.

**Comment by christf** [comment]
>  It illuminates the problem in K8s; there is no kube-proxy<->kubelet communication (and there shouldn't be!). They both react independently on kube-api updates.

I think this nails it. Because of this design, there is no way to first issue the updates to the ingress controller and only then handle the lifecycle. To fix it, these two processes need to be synchronized but that seems inherently incompatible with 
https://kubernetes.io/docs/concepts/architecture/controller/#controller-pattern

So, what if we extended the api for endpointSlices https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/? Currently  there is a field for serving, ready, terminating. These values are determined by the pod lifecycle.
could we have a field "activeIngress" or similar in there that is updated once the change to the ingress controller has been implemented?
If this was the case, then universally a preStop hook would work that could query this object and terminate, once the endpoint is inactive/properly drained. This would mean that the duration of this process would be subject to the TerminationGracePeriodSeconds. I feel,  this should be setting the upper bound for pod termination. If an endpoint cannot be drained and manually shut down then something outside of the scope of the k8s cluster is wrong.
This could even be implemented as part of kubelet - only relay the SIGTERM to the pod, once the field was updated....

Edit: maybe a status field would be helpful.

**Comment by karlkfi** [comment]
A data point from implementing a workaround for this, plus a smaller ask.

### On preemptible nodes, `preStop: sleep` makes things worse

`terminationGracePeriodSeconds` is a request, not a promise. During graceful node shutdown the kubelet gives a pod `min(TGPS, whatever the node has left)`. GKE enables graceful node shutdown by default with a 30s budget, split 15s ordinary / 15s critical. So an ordinary pod gets 15 seconds, whatever its spec says.

A `preStop: sleep 10` burns two thirds of that before the process is even signalled, leaving *less* time to drain than no hook at all. The advice in this thread is a sensible default, but it backfires exactly where pods get terminated most often.

### What works, with no Kubernetes change

**Wait inside the drain, not before it.** A `preStop` sleep runs before the shutdown code, so the grace period has to cover both in sequence. The two waits are unrelated and can overlap: in-flight work keeps finishing while endpoint removal propagates. Moving the wait into the SIGTERM handler makes the worst case `max(wait, drain)` instead of the sum, which is what keeps the grace period from having to grow.

**Then watch traffic instead of sleeping.** Keep the listener open after SIGTERM and track when a new connection last arrived. Once nothing new has come in for a couple of seconds, stop waiting. Measuring that interval from the later of (shutdown start, last arrival) gives a floor for an idle pod, and keeps extending while traffic is still landing. Cap it, since a quiet gap is evidence rather than proof.

The useful part: **you don't need to know who to wait for.** That's the question that has blocked the gate and webhook ideas above. Watching your own traffic covers every load balancer, ingress controller and kube-proxy at once, without naming or reaching any of them. It is weaker than a real gate (probabilistic, and it tells you nothing if the pod gets no traffic), but it beats a fixed sleep and needs nothing from upstream. It also differs from the earlier "watch your own endpoints" suggestion, which has to reason about state the pod cannot fully see.

### A smaller ask

If a termination gate is too big a change, here's a cheap one that would still help: **tell the pod how much grace period it actually got.**

The kubelet knows it is only giving this pod 15 seconds; the pod has no way to find out. Even the requested `terminationGracePeriodSeconds` isn't in the downward API, never mind the truncated value. With the real number, shutdown code could size itself to fit, and skip the endpoint wait entirely when there is no room for one.

That's SIG Node, where @thockin said most of this problem lives. No syncing, no knowing who routes traffic, no new lifecycle phase. Just surfacing a number the kubelet already has.

### Related: both of these reproduce on Kubernetes 1.35.0

Verified on kind. #116965 and #124648 compound, and between them nothing takes a pod out of rotation while it is shutting down on the graceful node shutdown path.

**#124648 (readiness probe stops too early at eviction).** With a readiness probe at 1s period and failureThreshold 1, starting to fail the instant SIGTERM arrives, the pod's Ready condition stayed True for the entire 48 second termination. It never flipped. On an ordinary `kubectl delete` this is masked, because the deletionTimestamp withdraws the endpoint anyway. It matters wherever that masking is absent.

**#116965 (graceful node shutdown does not update endpoints).** Kubelet configured `shutdownGracePeriod: 30s` / `shutdownGracePeriodCriticalPods: 15s`, two replicas behind a Service, node powered off:

| Time | Node | Pods | Endpoints |
|---|---|---|---|
| t+0 to t+14 | NotReady | Running, no deletionTimestamp | both still ready=true |
| t+16 | NotReady | Failed | gone |

The pods never entered Terminating and never received a deletionTimestamp. Their endpoints stayed ready for the full 15 second ordinary-pod budget, and were withdrawn only once the pods reached a terminal phase, which is after they had already stopped.

So on that path there is no deletionTimestamp to drive eager removal, and readiness is not observed either. Traffic keeps arriving for the whole window. That has two consequences for the workarounds discussed above: the traffic-watching approach cannot exit early there, and a lame-duck readiness endpoint does not help either, which is worth knowing given that readiness is the portable answer suggested on #116965.

One note for anyone reproducing this: stock kind cannot do graceful node shutdown at all. The node image ships no dbus, so the kubelet logs `Failed to start node shutdown manager: dial unix /var/run/dbus/system_bus_socket: no such file or directory` and the feature is silently inactive. Install dbus, start systemd-logind, restart the kubelet, and confirm `systemd-inhibit --list` shows the kubelet holding a shutdown lock before trusting any result.
